### PR TITLE
Stack 1426 - Refactor listener 

### PR DESCRIPTION
**Pull request by OmkarTelee-A10**
## Description
Refactoring of Listener flows and tasks with proper handling of reverts and certificate creation/update/delete tasks

## Jira Ticket
https://a10networks.atlassian.net/browse/STACK-1422
https://a10networks.atlassian.net/browse/STACK-1426
https://a10networks.atlassian.net/browse/STACK-1447

## Technical Approach
- Refactored relation of Pool to listener (switched to one:one) In Future we might consider one:many, but for now, its enough
- Refactored controller worker to allow single listener to be passed in flows
- Refactored certificate creation flow, added proper reverts and a linear flow making sure reverts happens smooth
- Refactored listener flows and tasks

## Config Changes
None

## Test case
All possible positive and revert test cases for listerner create/update/delete with various configuration

## Manual Testing

openstack loadbalancer listener create --protocol TCP --protocol-port 80 slb1
 openstack loadbalancer listener create --protocol HTTP --protocol-port 81 slb1
 openstack loadbalancer listener create --protocol HTTPS --protocol-port 82 slb1
 openstack loadbalancer listener create --protocol HTTPS --protocol-port 443 slb1
 openstack loadbalancer listener create --protocol TERMINATED_HTTPS --default-tls-container-ref http://10.64.3.108/key-manager/v1/containers/8ee2794e-9a3f-4f07-b5d4-48b1f91d1186 --protocol-port 90 slb1
 openstack loadbalancer listener create --protocol UDP --protocol-port 55 slb1 --connection-limit  30
 
 
 ###update test cases
 
 openstack loadbalancer listener set --connection-limit 40 a7144cdc-cd19-42a6-8670-9853e19b5a0b
 openstack loadbalancer listener set --default-tls-container-ref http://10.64.3.108/key-manager/v1/containers/f7edbe6e-201b-4105-a959-73d495c06af8 7c9fa388-18ef-48d2-ba4a-dc11634da47c
 
 
 ###delete test cases
 openstack loadbalancer listener delete 62404a7a-6ee6-4438-aebb-147e154a2ed6
 openstack loadbalancer listener delete 143b9028-6c4c-413b-8990-e5c1e814774f
 openstack loadbalancer listener delete 7c9fa388-18ef-48d2-ba4a-dc11634da47c
 openstack loadbalancer listener delete 4a3ecdd0-df79-4f50-a8fb-52afa8d09f82
 openstack loadbalancer listener delete 68e19cd8-fe96-4506-b8f4-543bd4f262e3
 openstack loadbalancer listener delete 2dd76eaf-408f-4067-bd49-fd70551b1ee2
 openstack loadbalancer listener delete a7144cdc-cd19-42a6-8670-9853e19b5a0b
 
 ####Pool create
 openstack loadbalancer pool create --protocol TCP --listener e314f853-0f47-46bb-97aa-822ac3bb93a5 --lb-algorithm LEAST_CONNECTIONS
 openstack loadbalancer pool delete 97d9e84b-dd9c-4420-b92b-626c8c180c36

### CODE REVIEW

File: a10_octavia/controller/worker/tasks/cert_tasks.py [hunk 1]
```diff
@@ -24,182 +24,200 @@
...
+class GetSSLCertData(task.Task):
+    """Task to get SSL certificate data from Barbican"""
+
+    def execute(self, loadbalancer, listener):
+        cert_data = None
+        if listener.protocol == 'TERMINATED_HTTPS' and listener.tls_certificate_id:
+            try:
+                barbican_client = BarbicanACLAuth().get_barbican_client(loadbalancer.project_id)
+                cert_data = utils.get_cert_data(barbican_client, listener)
+                # Monkey patching template name as listener ID
```
Comment (hthompson-a10) [inline a10_octavia/controller/worker/tasks/cert_tasks.py:36]: This isn't really monkey patching. You aren't replacing/adding a class function or anything.

**Comment by hthompson-a10** [comment]
https://docs.openstack.org/api-ref/load-balancer/v2/?expanded=update-a-listener-detail,create-a-load-balancer-detail#create-a-load-balancer

Seems like we are putting the final nail in the coffin of batch listener actions. Care needs to be taken regarding support for features lent by the Octavia API.

I've created a bug report for this in our backlog. Going to approve this refactor since the operation didn't work before this PR was submitted.

File: a10_octavia/controller/worker/tasks/cert_tasks.py [hunk 2]
```diff
@@ -24,182 +24,200 @@
...
+                                                           size=len(cert_data.cert_content),
+                                                           action="import",
+                                                           certificate_type=certificate_type)
+                else:
+                    self.axapi_client.file.ssl_cert.create(file=cert_data.cert_filename,
+                                                           cert=cert_data.cert_content,
+                                                           size=len(cert_data.cert_content),
+                                                           action="import",
+                                                           certificate_type=certificate_type)
+            except Exception as e:
```
Comment (hthompson-a10) [inline a10_octavia/controller/worker/tasks/cert_tasks.py:62]: We don't need to do a catch all exceptions here. That will obscure errors. There are only 2 that can be the issue:
1. `ACOSException` from `acos_client`
2. `ConnectTimeout` from `requests library`

We should attempt to catch both and in each case raise the error with the id of the listener/vthunder that is causing the failure.

This modification to exceptions should be complete across the entire refactor when dealing with only the `axapi_client`.

Openstack libraries can raise a larger number of errors which we don't have full visibility of so we can catch-all for them after trying to catch our known errors.

File: a10_octavia/controller/worker/tasks/cert_tasks.py [hunk 3]
```diff
@@ -102,40 +138,45 @@ def execute(self, cert_data, vthunder):
...
     def revert(self, cert_data, vthunder, *args, **kwargs):
         try:
+            LOG.warning("Reverting creation of SSL template: %s", cert_data.template_name)
             self.axapi_client.slb.template.client_ssl.delete(name=cert_data.template_name)
-            LOG.debug("Reverted ClientSSLTemplateCreate: %s", cert_data.template_name)
-        except Exception:
-            pass
+        except ConnectionError:
+            LOG.exception(
+                "Failed to connect A10 Thunder device: %s", vthunder.ip)
```
Comment (afrin-chakure-a10) [inline a10_octavia/controller/worker/tasks/cert_tasks.py:153]: This should be `vthunder.ip_address`

File: a10_octavia/controller/worker/tasks/virtual_port_tasks.py [hunk 4]
```diff
@@ -30,180 +29,130 @@
...
+        LOG.warning("Reverting creation of listener: %s", listener.id)
+        listener.protocol = openstack_mappings.virtual_port_protocol(self.axapi_client,
+                                                                     listener.protocol)
+        try:
+            self.axapi_client.slb.virtual_server.vport.delete(
+                loadbalancer.id, listener.id, listener.protocol,
+                listener.protocol_port)
+        except ConnectionError:
+            LOG.exception(
+                "Failed to connect A10 Thunder device: %s", vthunder.ip)
```
Comment (afrin-chakure-a10) [inline a10_octavia/controller/worker/tasks/virtual_port_tasks.py:124]: Should be `vthunder.ip_address`

**Pull request merged**